File: backend/services/file_parser.py
import io
import pdfplumber
from docx import Document


def extract_text_from_file(file_bytes: bytes, filename: str) -> str:
    """Extract plain text from an uploaded PDF or DOCX file."""
    ext = filename.lower().split(".")[-1]

    if ext == "pdf":
        return _extract_from_pdf(file_bytes)
    elif ext in ("docx", "doc"):
        return _extract_from_docx(file_bytes)
    else:
        raise ValueError(f"Unsupported file type: .{ext}. Please upload a PDF or DOCX file.")


def _extract_from_pdf(file_bytes: bytes) -> str:
    text_parts = []
    with pdfplumber.open(io.BytesIO(file_bytes)) as pdf:
        for page in pdf.pages:
            page_text = page.extract_text()
            if page_text:
                text_parts.append(page_text.strip())
    full_text = "\n\n".join(text_parts)
    if not full_text.strip():
        raise ValueError("Could not extract text from the PDF. It may be scanned/image-based.")
    return full_text


def _extract_from_docx(file_bytes: bytes) -> str:
    doc = Document(io.BytesIO(file_bytes))
    paragraphs = [p.text for p in doc.paragraphs if p.text.strip()]
    full_text = "\n".join(paragraphs)
    if not full_text.strip():
        raise ValueError("Could not extract text from the DOCX file.")
    return full_text

File: frontend/src/lib/file-parser.js
import * as pdfjsLib from 'pdfjs-dist';
import mammoth from 'mammoth';

// Point to the pdfjs worker (Vite will bundle it)
pdfjsLib.GlobalWorkerOptions.workerSrc = new URL(
  'pdfjs-dist/build/pdf.worker.mjs',
  import.meta.url
).href;

/**
 * Extract plain text from a File object (PDF or DOCX).
 */
export async function extractTextFromFile(file) {
  const ext = file.name.split('.').pop().toLowerCase();

  if (ext === 'pdf') {
    return await extractFromPdf(file);
  } else if (ext === 'docx' || ext === 'doc') {
    return await extractFromDocx(file);
  } else {
    throw new Error(`Unsupported file type: .${ext}. Please upload a PDF or DOCX file.`);
  }
}

async function extractFromPdf(file) {
  const arrayBuffer = await file.arrayBuffer();
  const pdf = await pdfjsLib.getDocument({ data: arrayBuffer }).promise;
  const textParts = [];

  for (let i = 1; i <= pdf.numPages; i++) {
    const page = await pdf.getPage(i);
    const content = await page.getTextContent();
    const pageText = content.items.map(item => item.str).join(' ');
    if (pageText.trim()) textParts.push(pageText.trim());
  }

  const fullText = textParts.join('\n\n');
  if (!fullText.trim()) {
    throw new Error('Could not extract text from the PDF. It may be a scanned image — please use a text-based PDF.');
  }
  return fullText;
}

async function extractFromDocx(file) {
  const arrayBuffer = await file.arrayBuffer();
  const result = await mammoth.extractRawText({ arrayBuffer });
  const text = result.value;
  if (!text.trim()) {
    throw new Error('Could not extract text from the DOCX file.');
  }
  return text;
}
